### PR TITLE
Performance updates

### PR DESCRIPTION
**Pull request by ColinEberhardt**
relates to #16 (but does not fix, still need to explore quad tree)

Both greedy and annealing iteratively move individual rectangles assessing the overall 'penalty score' for the layout. Currently they re-compute the entire layout penalty on each iteration.

This update instead computes the change in layout penalty for just the rectangle being moved. This has been refactored into a simple `rectangles` component (non public)

Also removes the `minimum` function, exchanging for d3-array `scan` as discussed in #18 

The changes to the `greedySpec` are because the previous implementation effected the order of the array by pushing rectangles to the top of the array. This optimisation doesn't change the array order.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,28 +23,32 @@ The label layout component provides a mechanism for arranging child components b
 ## Example usage
 
 ```javascript
+import { layoutTextLabel, layoutGreedy,
+    layoutLabel, layoutRemoveOverlaps } from 'd3fc-label-layout';
+import { select } from 'd3-selection';
+
 var labelPadding = 2;
```
Comment (chrisprice) [inline README.md]: `const`?

File: README.md [hunk 2]
```diff
@@ -23,28 +23,32 @@ The label layout component provides a mechanism for arranging child components b
...
+var strategy = layoutRemoveOverlaps(layoutGreedy());
 
 // create the layout that positions the labels
-var labels = fc.label(strategy)
-    .size(function(d) {
+var labels = layoutLabel(strategy)
+    .size((d, i, g) => {
         // measure the label and add the required padding
-        var textSize = d3.select(this)
+        var textSize = select(g[i])
```
Comment (chrisprice) [inline README.md]: Just use DOM methods?

File: README.md [hunk 3]
```diff
@@ -103,37 +107,37 @@ The strategy should return an array of objects indicating the placement of each
 
 The greedy strategy is a very fast way of reducing label overlap. It adds each label in sequence, selecting the position where the label has the lowest overlap with already added rectangles and is inside the container.
 
-*fc*.**greedy**()
+*fc*.**layoutGreedy**()
 
 Constructs a greedy strategy.
 
-*greedy*.**bounds**(*array*)
+*layoutGreedy*.**bounds**(*array*)
```
Comment (chrisprice) [inline README.md]: The comment indicates that the argument isn't an array?

File: README.md [hunk 4]
```diff
@@ -103,37 +107,37 @@ The strategy should return an array of objects indicating the placement of each
...
 
 The simulated annealing strategy runs over a set number of iterations, choosing a different location for one label on each iteration. If that location results in a better result, it is saved for the next iteration. Otherwise, it is saved with probability inversely proportional with the iteration it is currently on. This helps it break out of local optimums, hopefully producing better output. Because of the random nature of the algorithm, it produces variable output.
 
-*fc*.**annealing**()
+*fc*.**layoutAnnealing**()
 
 Constructs an annealing strategy.
 
-*annealing*.**bounds**(*array*)
+*layoutAnnealing*.**bounds**(*array*)
```
Comment (chrisprice) [inline README.md]: The comment indicates that the argument isn't an array?

File: index.js [hunk 5]
```diff
@@ -4,4 +4,4 @@ export { default as layoutGreedy } from './src/greedy';
 export { default as layoutAnnealing } from './src/annealing';
 export { default as layoutRemoveOverlaps } from './src/removeOverlaps';
 export { default as layoutBoundingBox } from './src/boundingBox';
-export { default as layoutIntersect } from './src/intersect';
+export { default as layoutIntersect } from './src/util/intersect';
```
Comment (chrisprice) [inline index.js]: Is it necessary to expose this?

File: src/util/layout.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        ...array.slice(index + 1)
+    ];
+
+const lessThan = (a, b) => a < b;
+
+// a layout takes an array of rectangles and allows their locations to be optimised.
+// it is constructed using two functions, locationScore, which score the placement of and
+// individual rectangle, and winningScore which takes the scores for a rectangle
+// at two different locations and assigns a winningScore.
+const layout = (rectangles, locationScore, winningScore, score) => {
```
Comment (chrisprice) [inline src/util/layout.js]: In a functional world, I would expect the less variable arguments to come first e.g. `(locationScore, winningScore, rectangles, score)`

Comment (chrisprice) [inline src/util/layout.js]: Any reason for this not being a component?

File: src/util/layout.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+    score = score || sum(
+        rectangles.map((_, i) => locationScore(rectangles, i))
+    );
+
+    winningScore = winningScore || lessThan;
+
+    const evaluatePlacement = (placement, index) =>
+        score -
+        locationScore(rectangles, index) +
```
Comment (chrisprice) [inline src/util/layout.js]: Could the arguments be changed to follow the d3/js convention of `rectangle, index, rectangles`?

File: src/util/layout.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        locationScore(substitute(rectangles, index, placement), index);
+
+    const tryLocation = (rectangle, index) => {
+        const newScore = evaluatePlacement(rectangle, index);
+        return winningScore(newScore, score)
```
Comment (chrisprice) [inline src/util/layout.js]: Another argument switch... `winningScore` is basically a reducer, could it take `current, new`?

File: src/util/layout.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            ? layout(substitute(rectangles, index, rectangle), locationScore, winningScore, newScore)
```
Comment (chrisprice) [inline src/util/layout.js]: I know you're not going for efficiency here but you're going to end up doing `substitute` twice per winning placement

File: src/util/layout.js (hunk 8 above)
Comment (chrisprice) [inline src/util/layout.js]: Actually why does this component know anything about `winningScore`?

File: test/greedySpec.js [hunk 10]
```diff
@@ -81,25 +81,25 @@ describe('greedy', () => {
         });
 
         it('moves a label half up, left if there is no prior fit', () => {
-            var sampledData = [data[1], data[3]];
```
Comment (chrisprice) [inline test/greedySpec.js]: These tests seem equivalent but just curious why you changed the order?

File: src/util/layout.js (hunk 6 above)
Comment (ColinEberhardt) [inline src/util/layout.js]: perhaps have the component return score & rectangles, rather than a new component

File: test/greedySpec.js (hunk 10 above)
Comment (ColinEberhardt) [inline test/greedySpec.js]: The original implementation of greedy changed the order of the rectangles in the layout. This implementation changes their location in-place.

File: src/util/layout.js (hunk 8 above)
Comment (ColinEberhardt) [inline src/util/layout.js]: winning score is needed for simulated annealing

**Comment by ColinEberhardt** [comment]
Thanks for the review comments, finally got round to addressing them. The `layout` logic works quite nicely as a component, and now has a clearer API.

**Pull request merged**